From 19bc056d83bcdd9b3938a9b7f8454bff4be39373 Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Thu, 4 May 2023 03:20:38 +0000
Subject: [PATCH 40/53] [Huawei] mm/sharepool: Reorder the procedure for adding
 a group

Offering: HULK
hulk inclusion
category: feature
bugzilla: N/A

--------------------------------

Move sp_group_setup_mapping() into group_add_task(). Because
the mapping for a new group is setup according to the task and
group_add_task() is indeed where we process both the task and
the group.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 mm/share_pool.c | 62 +++++++++++++++++++++++++++++--------------------
 1 file changed, 37 insertions(+), 25 deletions(-)

diff --git a/mm/share_pool.c b/mm/share_pool.c
index 571adf4e7b3d..e940345bc9a3 100644
--- a/mm/share_pool.c
+++ b/mm/share_pool.c
@@ -355,7 +355,7 @@ static struct sp_mapping *sp_mapping_create(unsigned long type)
 
 	spm = kzalloc(sizeof(struct sp_mapping), GFP_KERNEL);
 	if (!spm)
-		return ERR_PTR(-ENOMEM);
+		return NULL;
 
 	sp_mapping_set_type(spm, type);
 	sp_mapping_range_init(spm);
@@ -442,7 +442,7 @@ static bool can_mappings_merge(struct sp_mapping *m1, struct sp_mapping *m2)
  * the caller must hold sp_global_sem
  * NOTE: undo the mergeing when the later process failed.
  */
-static int sp_mapping_group_setup(struct mm_struct *mm, struct sp_group *spg)
+static int sp_group_setup_mapping_normal(struct mm_struct *mm, struct sp_group *spg)
 {
 	struct sp_mapping *local_dvpp_mapping, *spg_dvpp_mapping;
 
@@ -484,6 +484,35 @@ static int sp_mapping_group_setup(struct mm_struct *mm, struct sp_group *spg)
 	return 0;
 }
 
+static int sp_group_setup_mapping_local(struct mm_struct *mm, struct sp_group *spg)
+{
+	struct sp_mapping *spm;
+	struct sp_group *local = mm->sp_group_master->local;
+
+	spm = sp_mapping_create(SP_MAPPING_DVPP);
+	if (!spm)
+		return -ENOMEM;
+
+	sp_mapping_attach(local, spm);
+	sp_mapping_attach(local, sp_mapping_normal);
+	sp_mapping_attach(local, sp_mapping_ro);
+
+	return 0;
+}
+
+static inline bool is_local_group(int spg_id)
+{
+	return spg_id >= SPG_ID_LOCAL_MIN && spg_id <= SPG_ID_LOCAL_MAX;
+}
+
+static int sp_group_setup_mapping(struct mm_struct *mm, struct sp_group *spg)
+{
+	if (is_local_group(spg->id))
+		return sp_group_setup_mapping_local(mm, spg);
+	else
+		return sp_group_setup_mapping_normal(mm, spg);
+}
+
 static struct sp_group *sp_group_create(int spg_id);
 static void free_sp_group_locked(struct sp_group *spg);
 static int group_add_task(struct mm_struct *mm, struct sp_group *spg,
@@ -492,7 +521,6 @@ static int init_local_group(struct mm_struct *mm)
 {
 	int ret;
 	struct sp_group *spg;
-	struct sp_mapping *spm;
 	struct sp_group_master *master = mm->sp_group_master;
 
 	spg = sp_group_create(SPG_ID_LOCAL);
@@ -500,15 +528,6 @@ static int init_local_group(struct mm_struct *mm)
 		return PTR_ERR(spg);
 
 	master->local = spg;
-	spm = sp_mapping_create(SP_MAPPING_DVPP);
-	if (IS_ERR(spm)) {
-		ret = PTR_ERR(spm);
-		goto free_spg;
-	}
-	sp_mapping_attach(master->local, spm);
-	sp_mapping_attach(master->local, sp_mapping_normal);
-	sp_mapping_attach(master->local, sp_mapping_ro);
-
 	ret = group_add_task(mm, spg, PROT_READ | PROT_WRITE);
 	if (ret)
 		/* The spm would be released while destroying the spg */
@@ -560,11 +579,6 @@ static int sp_init_group_master_locked(struct task_struct *tsk, struct mm_struct
 	return ret;
 }
 
-static inline bool is_local_group(int spg_id)
-{
-	return spg_id >= SPG_ID_LOCAL_MIN && spg_id <= SPG_ID_LOCAL_MAX;
-}
-
 static struct sp_group *sp_get_local_group(struct task_struct *tsk, struct mm_struct *mm)
 {
 	int ret;
@@ -1268,6 +1282,10 @@ static int group_add_task(struct mm_struct *mm, struct sp_group *spg,
 	struct sp_group_node *node;
 	int ret;
 
+	ret = sp_group_setup_mapping(mm, spg);
+	if (ret)
+		return ret;
+
 	node = create_spg_node(mm, prot, spg);
 	if (IS_ERR(node))
 		return PTR_ERR(node);
@@ -1380,12 +1398,6 @@ int mg_sp_group_add_task(int tgid, unsigned long prot, int spg_id)
 		goto out_drop_group;
 	}
 
-	ret = sp_mapping_group_setup(mm, spg);
-	if (ret) {
-		up_write(&spg->rw_lock);
-		goto out_drop_group;
-	}
-
 	ret = group_add_task(mm, spg, prot);
 	if (ret) {
 		up_write(&spg->rw_lock);
@@ -3719,12 +3731,12 @@ static int __init share_pool_init(void)
 		return 0;
 
 	sp_mapping_normal = sp_mapping_create(SP_MAPPING_NORMAL);
-	if (IS_ERR(sp_mapping_normal))
+	if (!sp_mapping_normal)
 		goto fail;
 	atomic_inc(&sp_mapping_normal->user);
 
 	sp_mapping_ro = sp_mapping_create(SP_MAPPING_RO);
-	if (IS_ERR(sp_mapping_ro))
+	if (!sp_mapping_ro)
 		goto free_normal;
 	atomic_inc(&sp_mapping_ro->user);
 
-- 
2.17.1

